File: packages/app/headless/kevin/languages/go/internal/server/handler_test.go
package server

import (
	"strings"
	"testing"

	"github.com/hieudoanm/kevin/internal/db"
)

func TestHandleLine(t *testing.T) {
	tests := []struct {
		name  string
		setup func(*db.DB)
		line  string
		want  string
		reply bool
	}{
		{name: "ping", line: "PING", want: "PONG\n", reply: true},
		{name: "ping case-insensitive", line: "ping", want: "PONG\n", reply: true},
		{name: "set", line: "SET mykey myvalue", want: "OK\n", reply: true},
		{name: "set no args", line: "SET", want: "ERR usage: SET key value\n", reply: true},
		{name: "set one arg", line: "SET key", want: "ERR usage: SET key value\n", reply: true},
		{name: "get", setup: func(k *db.DB) { k.Set("mykey", "myvalue") }, line: "GET mykey", want: "myvalue\n", reply: true},
		{name: "get missing", line: "GET nonexistent", want: "(nil)\n", reply: true},
		{name: "get no args", line: "GET", want: "ERR usage: GET key\n", reply: true},
		{name: "get too many args", line: "GET a b", want: "ERR usage: GET key\n", reply: true},
		{name: "get trailing space ok", setup: func(k *db.DB) { k.Set("mykey", "myvalue") }, line: "GET mykey ", want: "myvalue\n", reply: true},
		{name: "del", setup: func(k *db.DB) { k.Set("mykey", "myvalue") }, line: "DEL mykey", want: "1\n", reply: true},
		{name: "del missing", line: "DEL nonexistent", want: "0\n", reply: true},
		{name: "del no args", line: "DEL", want: "ERR usage: DEL key\n", reply: true},
		{name: "del too many args", line: "DEL a b", want: "ERR usage: DEL key\n", reply: true},
		{name: "unknown command", line: "FOO", want: "ERR unknown command\n", reply: true},
		{name: "empty line", line: "", want: "", reply: false},
		{name: "whitespace line", line: "   ", want: "", reply: false},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			kv := db.New()
			if tt.setup != nil {
				tt.setup(kv)
			}
			got, ok := handleLine(tt.line, kv)
			if ok != tt.reply || got != tt.want {
				t.Errorf("handleLine(%q) = %q, %v; want %q, %v", tt.line, got, ok, tt.want, tt.reply)
			}
		})
	}
}

func TestValueWithSpaces(t *testing.T) {
	kv := db.New()
	if got, _ := handleLine("SET mykey hello world", kv); got != "OK\n" {
		t.Fatalf("SET with spaces = %q; want OK", got)
	}
	if got, _ := handleLine("GET mykey", kv); got != "hello world\n" {
		t.Errorf("GET after SET with spaces = %q; want %q", got, "hello world\n")
	}
}

func TestMultipleCommands(t *testing.T) {
	kv := db.New()
	seq := []struct{ cmd, want string }{
		{"SET a 1", "OK\n"},
		{"SET b 2", "OK\n"},
		{"GET a", "1\n"},
		{"GET b", "2\n"},
		{"DEL a", "1\n"},
		{"GET a", "(nil)\n"},
	}
	for _, s := range seq {
		if got, _ := handleLine(s.cmd, kv); got != s.want {
			t.Errorf("handleLine(%q) = %q; want %q", s.cmd, got, s.want)
		}
	}
}

func TestKeys(t *testing.T) {
	kv := db.New()
	kv.Set("a", "1")
	kv.Set("b", "2")
	kv.Set("c", "3")
	got, ok := handleLine("KEYS", kv)
	if !ok {
		t.Fatal("KEYS should produce a reply")
	}
	if got == "\n" {
		t.Fatal("KEYS should not be empty")
	}
	if count := len(strings.Fields(got)); count != 3 {
		t.Errorf("KEYS = %q; want 3 keys, got %d", got, count)
	}
}

func TestKeysEmpty(t *testing.T) {
	kv := db.New()
	if got, _ := handleLine("KEYS", kv); got != "\n" {
		t.Errorf("KEYS on empty DB = %q; want %q", got, "\n")
	}
}

File: packages/app/headless/kevin/languages/kotlin/src/test/kotlin/io/github/hieudoanm/kevin/server/HandlerTest.kt
package io.github.hieudoanm.kevin.server

import io.github.hieudoanm.kevin.db.Db
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

class HandlerTest {
    private var clock = 0L
    private val db = Db { clock }

    private fun run(line: String) = handleLine(line, db)

    @Test
    fun `ping replies pong`() {
        assertEquals(Reply("PONG\n", true), run("PING"))
    }

    @Test
    fun `commands are case insensitive`() {
        assertEquals(Reply("PONG\n", true), run("ping"))
        assertEquals(Reply("OK\n", true), run("sEt Foo Bar"))
        assertEquals(Reply("Bar\n", true), run("get foo"))
    }

    @Test
    fun `blank lines produce no reply`() {
        assertFalse(run("").shouldReply)
        assertFalse(run("   ").shouldReply)
        assertFalse(run("\r\n").shouldReply)
    }

    @Test
    fun `trailing carriage returns are trimmed`() {
        assertEquals(Reply("PONG\n", true), run("PING\r\n"))
    }

    @Test
    fun `unknown commands report an error`() {
        assertEquals(Reply("ERR unknown command\n", true), run("FROBNICATE"))
    }

    @Test
    fun `set and get a value`() {
        assertEquals(Reply("OK\n", true), run("SET name kevin"))
        assertEquals(Reply("kevin\n", true), run("GET name"))
    }

    @Test
    fun `set keeps embedded spaces in the value`() {
        run("SET greeting hello there world")
        assertEquals("hello there world", db.get("greeting"))
    }

    @Test
    fun `set without a value reports usage`() {
        assertEquals(Reply("ERR usage: SET key value [EX seconds]\n", true), run("SET"))
        assertEquals(Reply("ERR usage: SET key value [EX seconds]\n", true), run("SET onlykey"))
    }

    @Test
    fun `set with a valid ex stores a value without the suffix`() {
        assertEquals(Reply("OK\n", true), run("SET temp value EX 30"))
        assertEquals("value", db.get("temp"))
        assertEquals(30, db.ttl("temp").remaining.inWholeSeconds)
    }

    @Test
    fun `set with a non-positive or invalid ex reports an error`() {
        assertEquals(Reply("ERR invalid expire time\n", true), run("SET a b EX 0"))
        assertEquals(Reply("ERR invalid expire time\n", true), run("SET a b EX -5"))
        assertEquals(Reply("ERR invalid expire time\n", true), run("SET a b EX soon"))
        assertNull(db.get("a"))
    }

    @Test
    fun `get on a missing key replies nil`() {
        assertEquals(Reply("(nil)\n", true), run("GET ghost"))
    }

    @Test
    fun `get with extra arguments reports usage`() {
        assertEquals(Reply("ERR usage: GET key\n", true), run("GET a b"))
    }

    @Test
    fun `del removes a single key and counts it`() {
        run("SET a 1")
        assertEquals(Reply("1\n", true), run("DEL a"))
        assertEquals(Reply("0\n", true), run("DEL a"))
    }

    @Test
    fun `del accepts several keys`() {
        run("SET a 1")
        run("SET b 2")
        assertEquals(Reply("2\n", true), run("DEL a b"))
    }

    @Test
    fun `del without arguments reports usage`() {
        assertEquals(Reply("ERR usage: DEL key [key ...]\n", true), run("DEL"))
    }

    @Test
    fun `keys lists keys space separated`() {
        run("SET a 1")
        run("SET b 2")
        assertEquals(Reply("a b\n", true), run("KEYS"))
    }

    @Test
    fun `exists reports presence`() {
        run("SET a 1")
        assertEquals(Reply("1\n", true), run("EXISTS a"))
        assertEquals(Reply("0\n", true), run("EXISTS b"))
    }

    @Test
    fun `exists with extra arguments reports usage`() {
        assertEquals(Reply("ERR usage: EXISTS key\n", true), run("EXISTS a b"))
    }

    @Test
    fun `len reports the key count`() {
        run("SET a 1")
        run("SET b 2")
        assertEquals(Reply("2\n", true), run("LEN"))
    }

    @Test
    fun `len with arguments reports usage`() {
        assertEquals(Reply("ERR usage: LEN\n", true), run("LEN now"))
    }

    @Test
    fun `flushall and flushdb empty the store`() {
        run("SET a 1")
        assertEquals(Reply("OK\n", true), run("FLUSHALL"))
        assertEquals(Reply("0\n", true), run("LEN"))
        run("SET b 2")
        assertEquals(Reply("OK\n", true), run("FLUSHDB"))
        assertEquals(Reply("0\n", true), run("LEN"))
    }

    @Test
    fun `flush with arguments reports usage`() {
        assertEquals(Reply("ERR usage: FLUSHALL\n", true), run("FLUSHALL now"))
    }

    @Test
    fun `commands share one store across connections`() {
        run("SET shared 1")
        assertEquals(Reply("1\n", true), handleLine("EXISTS shared", db))
        assertTrue(db.exists("shared"))
    }
}
